warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,120 @@
+    user = login_service.get_user(userid)
+
+    if user is None:
+        return
+
+    if request.session.password_outdated(login_service.get_password_timestamp(userid)):
+        request.session.invalidate()
+        request.session.flash("Session invalidated by password change", queue="error")
+        return
+
+    principals = []
+
+    if user.is_superuser:
+        principals.append("group:admins")
+    if user.is_moderator or user.is_superuser:
+        principals.append("group:moderators")
+    if user.is_psf_staff or user.is_superuser:
+        principals.append("group:psf_staff")
+
+    # user must have base admin access if any admin permission
+    if principals:
+        principals.append("group:with_admin_dashboard_access")
+
+    return principals
+
+
+@implementer(ISecurityPolicy)
+class MultiSecurityPolicy:
+    """
+    A wrapper for multiple Pyramid 2.0-style "security policies", which replace
+    Pyramid 1.0's separate AuthN and AuthZ APIs.
+
+    Security policies are checked in the order provided during initialization,
+    with the following semantics:
+
+    * `identity`: Selected from the first policy to return non-`None`
+    * `authenticated_userid`: Selected from the request identity, if present
+    * `forget`: Combined from all policies
+    * `remember`: Combined from all policies
+    * `permits`: Uses the AuthZ policy passed during initialization
+
+    These semantics mostly mirror those of `pyramid-multiauth`.
+    """
+
+    def __init__(self, policies, authz):
+        self._policies = policies
+        self._authz = authz
+
+    def identity(self, request):
+        for policy in self._policies:
+            if ident := policy.identity(request):
+                return ident
+
+        return None
+
+    def authenticated_userid(self, request):
+        if request.identity:
+            return str(request.identity.id)
+        return None
+
+    def forget(self, request, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.forget(request, **kw))
+        return headers
+
+    def remember(self, request, userid, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.remember(request, userid, **kw))
+        return headers
+
+    def permits(self, request, context, permission):
+        identity = request.identity
+        principals = []
+        if identity is not None:
+            # Our identity is always an ORM model, so add some generic principals first.
+            principals.extend([Authenticated, str(identity.id)])
+
+            if isinstance(identity, User):
+                principals.extend(
+                    principals_for_authenticated_user(identity.id, request)
+                )
+            elif isinstance(identity, Project):
+                return Denied("unimplemented")
+
+        return self._authz.permits(context, principals, permission)
```

Comment:
Yeah, it's extremely confusing -- it's because we're really wrapping the old-style AuthZ APIs under the hood here. I'll add a comment that explains the behavior a bit.